### PR TITLE
Remove device parameter from create_extended_attention_mask_for_decoder

### PR DESCRIPTION
**Pull request by pbelevich**
# What does this PR do?

This RP removes redundant `device` parameter from `create_extended_attention_mask_for_decoder` that may cause potential issues if passed `device` is not equal `attention_mask.device`, see line [`modeling_utils.py#L610`](https://github.com/huggingface/transformers/blob/6d90d76f5db344e333ec184cc6f414abe2aa6559/src/transformers/modeling_utils.py#L610). Explanation: tracing logic from line 610 to method signature: 
`causal_mask.device` == `attention_mask.device` => `seq_ids.device` == `attention_mask.device` => `device` == `attention_mask.device`

@michaelbenayoun 

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by michaelbenayoun** [comment]
This seems legit for me, pinging @LysandreJik, @sgugger and @ydshieh to comment on this.

**Comment by ydshieh** [comment]
LGTM, as it uses the device from the argument `attention_mask`.

https://github.com/huggingface/transformers/blob/5d59df5e880cba38b1f2aa69acb8e5db0d84841f/src/transformers/modeling_utils.py#L592-L594

**Comment by ydshieh** [comment]
> LGTM, as it uses the device from the argument `attention_mask`.
> 
> https://github.com/huggingface/transformers/blob/5d59df5e880cba38b1f2aa69acb8e5db0d84841f/src/transformers/modeling_utils.py#L592-L594

(Please wait  the approvals from sgugger or LysandreJik before merge 🙏 )

**Review by ydshieh — APPROVED** [review]

File: examples/research_projects/bert-loses-patience/pabee/modeling_pabee_bert.py [hunk 1]
```diff
@@ -152,7 +152,7 @@ def forward(
 
         # We can provide a self-attention mask of dimensions [batch_size, from_seq_length, to_seq_length]
         # ourselves in which case we just need to make it broadcastable to all heads.
-        extended_attention_mask: torch.Tensor = self.get_extended_attention_mask(attention_mask, input_shape, device)
+        extended_attention_mask: torch.Tensor = self.get_extended_attention_mask(attention_mask, input_shape)
```
Comment (sgugger) [inline examples/research_projects/bert-loses-patience/pabee/modeling_pabee_bert.py:155]: This example is pinned to Transformers == 3.5.1 so don't make any change there.

File: examples/research_projects/deebert/src/modeling_highway_bert.py [hunk 2]
```diff
@@ -195,7 +195,7 @@ def forward(
 
         # We can provide a self-attention mask of dimensions [batch_size, from_seq_length, to_seq_length]
         # ourselves in which case we just need to make it broadcastable to all heads.
-        extended_attention_mask: torch.Tensor = self.get_extended_attention_mask(attention_mask, input_shape, device)
+        extended_attention_mask: torch.Tensor = self.get_extended_attention_mask(attention_mask, input_shape)
```
Comment (sgugger) [inline examples/research_projects/deebert/src/modeling_highway_bert.py:198]: Same here.

Comment (sgugger) [inline examples/research_projects/deebert/src/modeling_highway_bert.py:198]: Same here.

File: examples/research_projects/longform-qa/eli5_utils.py [hunk 3]
```diff
@@ -137,7 +137,7 @@ def embed_sentences_checkpointed(self, input_ids, attention_mask, checkpoint_bat
             token_type_ids = torch.zeros(input_shape, dtype=torch.long, device=device)
             head_mask = [None] * self.sent_encoder.config.num_hidden_layers
             extended_attention_mask: torch.Tensor = self.sent_encoder.get_extended_attention_mask(
-                attention_mask, input_shape, device
+                attention_mask, input_shape
```
Comment (sgugger) [inline examples/research_projects/longform-qa/eli5_utils.py:140]: This one can be updated as it's not pinned.

Comment (sgugger) [inline examples/research_projects/longform-qa/eli5_utils.py:140]: This one can be updated as it's not pinned.

File: src/transformers/modeling_utils.py [hunk 4]
```diff
@@ -589,8 +589,9 @@ def invert_attention_mask(self, encoder_attention_mask: Tensor) -> Tensor:
 
         return encoder_extended_attention_mask
 
-    def create_extended_attention_mask_for_decoder(self, input_shape, attention_mask, device):
+    def create_extended_attention_mask_for_decoder(self, input_shape, attention_mask):
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:592]: Removing an argument from a public method like this is a breaking change, so we should continue to accept if with a default of `None` than raise a deprecation warning if we detect it's not `None` telling the user that argument is not used anymore and will be removed in v5 of Transformers.

Comment (sgugger) [inline src/transformers/modeling_utils.py:592]: Removing an argument from a public method like this is a breaking change, so we should continue to accept if with a default of `None` than raise a deprecation warning if we detect it's not `None` telling the user that argument is not used anymore and will be removed in v5 of Transformers.

File: src/transformers/modeling_utils.py [hunk 5]
```diff
@@ -610,7 +611,7 @@ def create_extended_attention_mask_for_decoder(self, input_shape, attention_mask
         extended_attention_mask = causal_mask[:, None, :, :] * attention_mask[:, None, None, :]
         return extended_attention_mask
 
-    def get_extended_attention_mask(self, attention_mask: Tensor, input_shape: Tuple[int], device: device) -> Tensor:
+    def get_extended_attention_mask(self, attention_mask: Tensor, input_shape: Tuple[int]) -> Tensor:
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:614]: Same here.

Comment (sgugger) [inline src/transformers/modeling_utils.py:614]: Same here.

**Review by sgugger — APPROVED** [review]
Thanks for your PR. Removing arguments from public methods is a bit of a breaking change (even if I don't expect many users to use those directly) so since it's very easy to avoid it here and raise a proper deprecation warning, I would like to this added before we merge.

Also, the first two research projects should not be touched.

File: src/transformers/modeling_utils.py [hunk 6]
```diff
@@ -594,7 +594,11 @@ def invert_attention_mask(self, encoder_attention_mask: Tensor) -> Tensor:
         return encoder_extended_attention_mask
 
     @staticmethod
-    def create_extended_attention_mask_for_decoder(input_shape, attention_mask, device):
+    def create_extended_attention_mask_for_decoder(self, input_shape, attention_mask, device=None):
+        if device is not None:
+            warnings.warn("`device` is deprecated and will be removed in v5 of Transformers.")
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:599]:
```suggestion
            warnings.warn("The `device` argument is deprecated and will be removed in v5 of Transformers.", FutureWarning)
```

File: src/transformers/modeling_utils.py [hunk 7]
```diff
@@ -625,11 +631,16 @@ def get_extended_attention_mask(self, attention_mask: Tensor, input_shape: Tuple
             input_shape (`Tuple[int]`):
                 The shape of the input to the model.
             device: (`torch.device`):
+                **DEPRECATED**. `attention_mask.device` will be used instead in v5 of Transformers.
                 The device of the input to the model.
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:635]: Remove the documentation entirely for a deprecated argument.

File: src/transformers/modeling_utils.py [hunk 8]
```diff
@@ -625,11 +631,16 @@ def get_extended_attention_mask(self, attention_mask: Tensor, input_shape: Tuple
...
 
         Returns:
             `torch.Tensor` The extended attention mask, with a the same dtype as `attention_mask.dtype`.
         """
+        if not (attention_mask.dim() == 2 and self.config.is_decoder):
+            # show warning only if it won't be shown in `create_extended_attention_mask_for_decoder`
+            if device is not None:
+                warnings.warn("`device` is deprecated and will be removed in v5 of Transformers.")
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:643]:
```suggestion
                warnings.warn("The `device` argument is deprecated and will be removed in v5 of Transformers.", FutureWarning)
```

**Comment by pbelevich** [comment]
@sgugger thanks for the code review! all comments have been addressed

**Comment by sgugger** [comment]
Thanks! Pinging @LysandreJik for final review :-)

**Review by LysandreJik — APPROVED** [review]
LGTM, thanks @pbelevich!
